lib/autoproj/daemon/github_watcher.rb

```diff
@@ -132,141 +186,109 @@ def call_pull_request_hooks(pull_request_event)
 
-                push_events.select! { |e| e.owner == owner }
-                pull_request_events.select! { |e| e.pull_request.base_owner == owner }
+                modified_pull_requests = process_modified_pull_requests(events)
+                modified_mainlines = process_modified_mainlines(events)
 
-                handle_push_events(push_events)
-                handle_pull_request_events(pull_request_events)
+                dispatch(modified_mainlines, modified_pull_requests)
             end
 
-            # Compute the actual push events and dispatch them
+            # @api private
             #
-            # Instead of passing along the push events, we use them as hints
-            # that something happened on the branch we watch. We then pass along
-            # PushEvent objects where the SHA is the actual SHA from the branch,
-            # as reported by the API. This works arounds missing/wrong events
+            # Extracts all pull requests that might have been modified by the
+            # given events
             #
-            # This method does the generation of these 'events' based on the
-            # push events received from GitHub, and dispatches them
-            #
-            # @param [Array<Github::PushEvent>]
-            # @return [void]
-            def handle_push_events(events)
-                memo = {}
-                events = events.group_by(&:name)
-                events.each_value do |repo_events|
-                    repo_events = repo_events.group_by(&:branch)
-                    repo_events.each_value do |branch_events|
-                        latest_event = branch_events.max_by(&:created_at)
-                        sha = branch_current_head(
-                            latest_event.owner, latest_event.name, latest_event.branch,
-                            memo: memo
+            # @param [PartitionedEvents] events
+            # @return [GitHub::PullRequest=>[GitHub::PullRequestEvent,GitHub::PushEvent]]
+            def process_modified_pull_requests(events)
+                push_events =
+                    events
+                    .push_events_to_pull_request
+                    .group_by do |e|
+                        cached_pull_request_affected_by_push_event(
+                            e.owner, e.name, e.branch
                         )
-                        next unless sha
-
-                        event = latest_event.dup
-                        event.head_sha = sha
-                        dispatch_push_event(event)
                     end
-                end
+                push_events.delete(nil)
+                pull_request_events =
+                    events
+                    .pull_request_events
+                    .group_by(&:pull_request)
+
+                push_events =
+                    push_events
+                    .transform_keys { |pr| [pr.base_owner, pr.base_name, pr.number] }
```

Comment:
`transform_keys` is not available on ruby < 2.5 (CI is failing because of that)

Reply:
Do you mind if I add [`backports`](https://rubygems.org/gems/backports) as a dependency ? It's a very popular (and useful) gem.

Reply:
Nope... You may also drop support for ruby < 2.5.0 if you prefer to, I think you are the only one using this. Just wanted to point out because the tests were failing